cmd/pebble/cmd_enter.go

```diff
@@ -0,0 +1,164 @@
+`
+
+type enterFlags int
+
+const (
+	enterFlagSilcenceLogging enterFlags = 1 << iota
+	enterFlagNoServiceManager
+	enterFlagKeepServiceManager
+	enterFlagRequireServiceAutostart
+	enterFlagProhibitServiceAutostart
+)
+
+type haveEnterFlags interface {
+	enterFlags() enterFlags
+}
+
+type cmdEnter struct {
+	clientMixin
+
+	sharedRunEnterOpts
+	Run        bool `long:"run"`
+	Positional struct {
+		Cmd []string `positional-arg-name:"<subcommand>"`
+	} `positional-args:"yes"`
+}
+
+func init() {
+	optsHelp := map[string]string{
+		"run": "Start services before executing subcommand",
+	}
+	for k, v := range sharedRunEnterOptsHelp {
+		optsHelp[k] = v
+	}
+	cmdInfo := addCommand("enter", shortEnterHelp, longEnterHelp, func() flags.Commander { return &cmdEnter{} }, optsHelp, nil)
+	cmdInfo.extra = func(cmd *flags.Command) {
+		cmd.PassAfterNonOption = true
+	}
+}
+
+func (cmd *cmdEnter) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	runCmd := cmdRun{
+		sharedRunEnterOpts: cmd.sharedRunEnterOpts,
+	}
+	runCmd.setClient(cmd.client)
+
+	if len(cmd.Positional.Cmd) == 0 {
+		runCmd.run(nil)
+		return nil
+	}
+
+	runCmd.Hold = !cmd.Run
+
+	enterCommands := make([]*cmdInfo, 0)
+	enterDebugCommands := make([]*cmdInfo, 0)
+
+	for _, info := range commands {
+		if info.supportedInEnter {
+			enterCommands = append(enterCommands, info)
+		}
+	}
+	for _, info := range debugCommands {
+		if info.supportedInEnter {
+			enterDebugCommands = append(enterDebugCommands, info)
+		}
+	}
+
+	var (
+		subcCommander flags.Commander
+		subcExtraArgs []string
+		enterFlags    enterFlags
+	)
+
+	subcParser := createParser(cmd.client, enterCommands, enterDebugCommands)
+	subcParser.CommandHandler = func(commander flags.Commander, args []string) error {
+		if withFlags, ok := commander.(haveEnterFlags); ok {
+			enterFlags = withFlags.enterFlags()
+		}
+		subcCommander = commander
+		subcExtraArgs = args
+		return nil
+	}
+
+	if _, err := subcParser.ParseArgs(cmd.Positional.Cmd); err != nil {
+		return err
+	}
+
+	if enterFlags&enterFlagRequireServiceAutostart != 0 && !cmd.Run {
+		return fmt.Errorf("this enter subcommand can only operate with autostarted services (hint: add --run to enter)")
+	}
+
+	if enterFlags&enterFlagProhibitServiceAutostart != 0 && cmd.Run {
+		return fmt.Errorf("this enter subcommand cannot operate with autostarted services (hint: remove --run from enter)")
+	}
+
+	if enterFlags&enterFlagNoServiceManager != 0 {
+		if cmd.Run {
+			return fmt.Errorf("this enter subcommand runs without service manager (hint: remove --run from enter)")
+		}
+		return subcCommander.Execute(subcExtraArgs)
+	}
+
+	if enterFlags&enterFlagSilcenceLogging != 0 {
+		// what if cmd.Verbose (-v) is set?
+		logger.SetLogger(logger.NullLogger)
+	}
+
+	runReadyCh := make(chan func(), 1)
+	runPanicCh := make(chan interface{})
+	var runStop func()
+	var runPanic interface{}
+
+	go func() {
+		defer func() { runPanicCh <- recover() }()
```

Comment:
Why don't we just let a panic crash the process?

Reply:
This is because in `cmd/` we actually use `panic({exitStatus(1)})` as kind of _long jump_. When `exitStatus` error is recovered in `main`, the main just exits with the given status code. The run command uses it for restart functionality and also to bail when something goes wrong. If I let the goroutine panic, the `exitStatus` error wouldn't be recovered in main. And, since `exitStatus` has `Error` method, if run command wanted to exit with status 1, then the following would be printed
```
panic: internal error: exitStatus{1} being handled as normal error

goroutine 6 [running]:
main.(*cmdRun).run(0x0?, 0x0?)
	/home/woky/devel/pebble/cmd/pebble/cmd_run.go:82 +0xa9
main.(*cmdEnter).Execute.func2()
	/home/woky/devel/pebble/cmd/pebble/cmd_enter.go:140 +0x25
created by main.(*cmdEnter).Execute
	/home/woky/devel/pebble/cmd/pebble/cmd_enter.go:138 +0x46a
exit status 2
```

instead of just

```
exit status 1
```